### ISSUE TITLE
JSON Facet API

### ISSUE THREAD

**Issue by jr-ewing** [solariumphp/solarium#378]
Is there a support for JSON Facet API for Solarium 3.x available ?

See http://yonik.com/json-facet-api/

$ curl http://localhost:8983/solr/query -d 'q=_:_&rows=0&
 json.facet={
   categories:{
     type : terms,
     field : cat,
     sort : { x : desc},
     facet:{
       x : "avg(price)",
       y : "sum(price)"
     }
   }
 }

Or can this be done by $query->addParam ?

best regards
Tom

**Comment by basdenooijer** [comment]
Solarium doesn't have an API for this yet, however it is possible to add a custom request param to your Solr request as a workaround. 
Support for the JSON api should be added to Solarium in the future.

**Comment by bpolaszek** [comment]
:+1: 

Moving to PHP >= 5.4 as a base requirement would allow Solarium facet building objects to implement PHP native `\JsonSerializable` interface. 
We could then build the whole JsonFacet with a simple json_encode() of the different nested facet objects.

I think it would be a nice feature to implement, as it would make facet building easier.

**Comment by basdenooijer** [comment]
I agree it would be better to support JSON for facets. For the next major version of Solarium the minimum PHP version is getting bumped up anyway (to at least 5.6) so that wouldn't be an issue.

However I'm not sure if the JsonSerializable interface should be used. Currently the facet implementations are split between a value object and a request mapper, by implementing the Json serialization in the object itself it gets multiple responsibilities. But, something similar can also be done in a separate mapper still using JSON. 

I would be quite some work though, as all currently implemented facet types including a lot of testcases would need refactoring.

**Comment by bpolaszek** [comment]
In a perfect world, an external serializer should indeed be used to transform value objects to a JSON representation.

The dark side is that it requires a lot of work, and should be part of the solarium library, otherwise it makes no sense.
Besides, serializing would not only be useful for creating requests, but also to serialize results. For instance, I currently make use of the solarium library in a production app to display some results, facets, groups, paging, filtering (showing applied facets), stats, etc. It's OK in a templating engine, since you have to code every single place where to display a property of a solarium object. But when it came to represent data in a REST API, I had to override every single value object from the library to make them JSON-friendly. Of course I could have coded my own serializer, but as I just wanted something simple (value objects as JSON, without renaming properties), it was like going to the bakery  with a truck instead of a bike (I'm not good at comparisons, but you've got the point). So I went to the easier way.

Using the `\JsonSerializable` interface on value objects doesn't really give them responsibilities (no "action" is needed inside the `jsonSerialize()`: you don't have to call any submethod), it's just a way to tell the app "here's what I look like if you `json_encode` me", the same way `__toString()` tells the app "here's what I look like if you `(string)` me".

Simple as that:
```php
class AnimalCollection implements \JsonSerializable {

    protected $animals = [];

    /**
     * AnimalCollection constructor.
     * @param array $animals
     */
    public function __construct(array $animals) {
        $this->animals = $animals;
    }

    /**
     * @inheritDoc
     */
    public function jsonSerialize() {
        return [
            'animals' => $this->animals,
        ];
    }
}

class Animal implements \JsonSerializable {

    protected $type;

    /**
     * Animal constructor.
     * @param $type
     */
    public function __construct($type) {
        $this->type = $type;
    }

    /**
     * @inheritDoc
     */
    public function __toString() {
        return (string) $this->type;
    }

    /**
     * @inheritDoc
     */
    public function jsonSerialize() {
        return (string) $this;
    }

}

print(json_encode(new AnimalCollection([new Animal('cat'), new Animal('dog')]), JSON_PRETTY_PRINT));
```
outputs:
```json

{
    "animals": [
        "cat",
        "dog"
    ]
}
```

This doesn't prevent the use of an external serializer if one need some more complex logic. And obviously this interface should be used on value objects only. 

The other point is that you expect it to work "out of the box". 
Tomorrow, if a Solarium contributor adds some more value classes, he'll have to tell the external serializer how to handle them. If you use the `\JsonSerializable` interface (or another interface that extends it), you already know it makes part of the same "basic" serialization process.

I think this debate is almost the same as using Doctrine annotations on entities. Some coders prefer using an external XML/YML config, some others prefer to define the object's representation in a database directly in the entity class... But the Doctrine library offers the two possibilities.

**Comment by rakeshgithub** [comment]
May I know how to pass the custom params in query? , i need to use facet.json in query

**Comment by jr-ewing** [comment]
is there not a simple workarround to add something like -> json.facet={x:'avg(price)'} to the Request ?

**Comment by rakeshgithub** [comment]
I just figured out it by customizerequest plugin of solarium, below is the code snippet.

Hope it will help.

`$client = new Solarium\Client($config_solr); 
$customizer = $client->getPlugin('customizerequest');
$facet_json = "{x:'avg(price)'}";
$sql_and = "your solr query";

        // add a GET param thats only used for a single request (the default setting is no persistence)
        $customizer->createCustomization('json.facet')
            ->setType('param')
            ->setName('json.facet')
            ->setValue($facet_json);

        // get a select query instance
        $query = $client->createSelect();
        $query->setQuery($sql_and);
        $solr_result = $client->select($query);`

**Comment by jr-ewing** [comment]
Thank you - i will try !

**Comment by rakeshgithub** [comment]
Its inbuilt in solarium, see http://solarium.readthedocs.io/en/stable/plugins/#customizerequest-plugin

**Comment by jr-ewing** [comment]
ahh great ! Thank you

**Comment by mkalkbrenner** [comment]
solved by PR #590

**Comment by wwpstl** [comment]
> I just figured out it by customizerequest plugin of solarium, below is the code snippet.
> 
> Hope it will help.
> 
> `$client = new Solarium\Client($config_solr);
> $customizer = $client->getPlugin('customizerequest');
> $facet_json = "{x:'avg(price)'}";
> $sql_and = "your solr query";
> 
> ```
>     // add a GET param thats only used for a single request (the default setting is no persistence)
>     $customizer->createCustomization('json.facet')
>         ->setType('param')
>         ->setName('json.facet')
>         ->setValue($facet_json);
> 
>     // get a select query instance
>     $query = $client->createSelect();
>     $query->setQuery($sql_and);
>     $solr_result = $client->select($query);`
> ```

thanks
